### PR TITLE
[COLLABORATION] Added logic to interpret file system events and save to file system

### PR DESCRIPTION
**Pull request by hkirat**
In reference to [#2342](https://github.com/mozilla/thimble.mozilla.org/issues/2342)

### CODE REVIEW

File: src/editor/Collaboration.js [hunk 1]
```diff
@@ -160,18 +166,136 @@ define(function (require, exports, module) {
...
+                result.reject(fileError);
+            });
+        return result.promise();
+    }
+
     function _getOpenCodemirrorInstance(fullPath) {
-        var masterEditor = EditorManager.getCurrentFullEditor();
-        if(masterEditor.getFile().fullPath === fullPath) {
-            return masterEditor._codeMirror;
+        var doc = DocumentManager.getOpenDocumentForPath(fullPath);
```
Comment (hkirat) [inline src/editor/Collaboration.js]: Shifted to syncing all open codemirror instances instantly rather than just the one open because the control doesn't reach [here](https://github.com/mozilla/brackets/blob/collaboration/src/document/DocumentManager.js#L337) for these files.
If a file has a codemirror instance attached to it, the control reaches [here](https://github.com/mozilla/brackets/blob/collaboration/src/document/DocumentManager.js#L333) and it doesn't read from the filesystem.

File: src/editor/Collaboration.js [hunk 2]
```diff
@@ -7,11 +7,16 @@ define(function (require, exports, module) {
     var Path            = require("filesystem/impls/filer/FilerUtils").Path;
     var EditorManager   = require("editor/EditorManager");
     var CommandManager  = require("command/CommandManager");
+    var DocumentManager = require("document/DocumentManager");
+    var FileUtils       = require("file/FileUtils");
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: you don't really need this. You can use the same function from `FilerUtils` instead

File: src/editor/Collaboration.js [hunk 3]
```diff
@@ -160,18 +166,136 @@ define(function (require, exports, module) {
...
+     *
+     * @param {!string} path
+     * @return {$.Promise} A promise object that will be resolved when the buffer array for the file
+     * is cleared or rejected with a FileSystemError if the file is not yet open and can't be read from disk.
+     */
+    function clearFile(path) {
+        var result = new $.Deferred();
+        var file = FileSystem.getFileForPath(path);
+        if(!_webrtc || !_buffer || !_buffer[path] || _buffer[path].length === 0) {
+            FileUtils.readAsText(file)
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: You're not doing anything special here so you don't need any of the promise logic. Instead, this should actually just be:
```js
return FilerUtils.readAsUTF8(file);
```

Comment (gideonthomas) [inline src/editor/Collaboration.js]: in fact you can probably move this all the way to the top of this function.

File: src/editor/Collaboration.js [hunk 4]
```diff
@@ -160,18 +166,136 @@ define(function (require, exports, module) {
...
+                .done(function (rawText, readTimestamp) {
+                    result.resolve(rawText, readTimestamp);    
+                })
+                .fail(function (fileError) {
+                    result.reject(fileError);
+                });
+            return result.promise();
+        }
+        FileUtils.readAsText(file)
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: always add blank lines between block-level statements.

File: src/editor/Collaboration.js [hunk 5]
```diff
@@ -160,18 +166,136 @@ define(function (require, exports, module) {
...
+
+    /**
+     * Applies all the changes kept in the buffer array that have occured on connected clients but have not
+     * yet been written to the filesystem for this file.
+     *
+     * @param {!string} path
+     * @return {$.Promise} A promise object that will be resolved when the buffer array for the file
+     * is cleared or rejected with a FileSystemError if the file is not yet open and can't be read from disk.
+     */
+    function clearFile(path) {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: Or maybe `updateFileWithDiffs`

File: src/document/DocumentManager.js [hunk 6]
```diff
@@ -361,11 +362,11 @@ define(function (require, exports, module) {
                     PerfUtils.finalizeMeasurement(perfTimerName);
                 });
 
-                FileUtils.readAsText(file)
+                Collaboration.applyDiffsToFile(fullPath)
                     .always(function () {
                         // document is no longer pending
                         delete getDocumentForPath._pendingDocumentPromises[file.id];
-                    })
+                    })
```
Comment (gideonthomas) [inline src/document/DocumentManager.js]: undo this change

File: src/document/DocumentManager.js [hunk 7]
```diff
@@ -377,7 +378,6 @@ define(function (require, exports, module) {
                     .fail(function (fileError) {
                         result.reject(fileError);
                     });
-
```
Comment (gideonthomas) [inline src/document/DocumentManager.js]: undo this change

File: src/editor/Collaboration.js [hunk 8]
```diff
@@ -133,7 +140,6 @@ define(function (require, exports, module) {
         }
         var fullPath = Path.join(StartupState.project("root"), relPath);
         var codemirror = _getOpenCodemirrorInstance(fullPath);
-
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: undo this change

**Comment by gideonthomas** [comment]
@hkirat would you be able to fix the conflicts? I can merge it then

**Comment by hkirat** [comment]
Fixed the merge conflicts!
